[android] Add libtorch headers to pytorch_android aar

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40176 [android] gradle version update
* #39587 [android] test_app example linking to pytorch_android aar content
* #39588 [android] ANDROID_STL=c++_shared
* **#39507 [android] Add libtorch headers to pytorch_android aar**
* #39584 [android][cmake] Remove NO_EXPORT for libtorch mobile build

Adding gradle task that will be run after `assemble` to add `headers` folder to the aar.

Headers are choosed for the first specified abi, they should be the same for all abis.

Adding headers works through temporary unpacking into gradle `$buildDir`, copying headers to it, zipping aar with headers.

Differential Revision: [D22118009](https://our.internmc.facebook.com/intern/diff/D22118009)

IvanKobzarev committed:
[android] Add libtorch headers to pytorch_android aar

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"




[ghstack-poisoned]

ljk53 commented on android/pytorch_android/build.gradle:
```diff
@@ -79,4 +79,49 @@ task sourcesJar(type: Jar) {
+              "addHeadersToAar" + variant.name,
+              f.path.replaceAll(variant.name, VERSION_NAME),
+              getLibtorchHeadersDir(),
+              "headers")
```
is there any convention about the naming of header directory in AAR?

IvanKobzarev replied:
I have not heard about any conventions except [prefab](https://google.github.io/prefab/) that has some structure, but it is not ready for publishing yet. Probably, in 1-2 months we will be able to publish prefab package, but that would be consumable only by android gradle plugin 4.2.0+, gradle 6.5.

In some other packages, like tensorflow lite aar, 'headers' is also used as path for headers inside aar.

ljk53 commented on android/pytorch_android/build.gradle:
```diff
@@ -79,4 +79,49 @@ task sourcesJar(type: Jar) {
+  }
+}
+
+def addFolderToAarTask(taskName, aarPath, folderPath, folderPathInAar) {
```
just curious: can we include header files before packing AAR file instead of unpacking/repacking here?

IvanKobzarev replied:
I tried approach for including headers during packing of aar, it was very hacky and dependent on android gradle plugin internal structure, that may change version to version. I decided that repacking aar will be more stable for updating android gradle plugin versions.

ljk53 commented on android/pytorch_android/build.gradle:
```diff
@@ -79,4 +79,49 @@ task sourcesJar(type: Jar) {
+    android.libraryVariants.all { variant ->
+      variant.outputs.each { output ->
+        File f = output.outputFile
+        if (f.name.endsWith(".aar")) {
```
I'm not familiar with `android.libraryVariants.all` above - is there supposedly only one AAR file?

IvanKobzarev replied:
for `pytorch_android` variants.all will include `debug` and `release`.

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"


Adding gradle task that will be run after `assemble` to add `headers` folder to the aar.

Headers are choosed for the first specified abi, they should be the same for all abis.

Adding headers works through temporary unpacking into gradle `$buildDir`, copying headers to it, zipping aar with headers.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"


Adding gradle task that will be run after `assemble` to add `headers` folder to the aar.

Headers are choosed for the first specified abi, they should be the same for all abis.

Adding headers works through temporary unpacking into gradle `$buildDir`, copying headers to it, zipping aar with headers.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"


Adding gradle task that will be run after `assemble` to add `headers` folder to the aar.

Headers are choosed for the first specified abi, they should be the same for all abis.

Adding headers works through temporary unpacking into gradle `$buildDir`, copying headers to it, zipping aar with headers.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] Add libtorch headers to pytorch_android aar"


Adding gradle task that will be run after `assemble` to add `headers` folder to the aar.

Headers are choosed for the first specified abi, they should be the same for all abis.

Adding headers works through temporary unpacking into gradle `$buildDir`, copying headers to it, zipping aar with headers.

Differential Revision: [D22118009](https://our.internmc.facebook.com/intern/diff/D22118009)

[ghstack-poisoned]
